### DISCUSSION TITLE
Edit web part settings through PnP/PowerShell

### DISCUSSION THREAD

**Discussion by Niax01** [microsoft-search/pnp-modern-search#2015] [Q&A] [answered]
Hi all,

Is it possible to programmatically add/change the search query template?
I am building about 600 pages with a script, where each page has 2 search results web parts and the query template for both these web parts needs to be changed on each page after that they have been created, tedious :)
I have all the information I need to build this query string while creating the page, would be nice if I also could add this automagically.
Maybe there is a way to access the Import function through PnP/PowerShell?

/Nicklas

**Comment by wobba** [comment, marked as answer]
Yes, you need to look at how the web part properties are serialized, and then modify it. You can for example use something like  `Set-PnPClientSideWebPart -Page MyPage -Identity <webpartid> -PropertiesJson <json>`

You can provision pages with the web part using the PnP Provisioning schema as well. And then use tokens to replace the query string as needed.

**Reply by Niax01** [reply]
Thanks a lot Mikael, this at least gave me a hint on how to move forward :)
But since I am not that familiar working with code I kind of get stuck on the -PropertiesJson part.
I have extracted it from the web part InstanceId, which gave me a really long "string" of all the settings where I guess this part is what I am looking for.
`"dataSourceProperties":{"queryTemplate":"{searchTerms}path:https://XXX.sharepoint.com/sites/Internalrules-int/SitePages FileType:(\"aspx\") AND RefinableString00:\"Human Resources\""`
But now I am stuck, not sure how to change/set "path" to what I want.
Or do I need to save it all as a variable and use Set-PnPClientSideWebPart?
How would I then first get an output of it, with correct formatting, so I could add it to my own variable?

Sorry for all the stupid questions, but I am learning :)

/Nicklas

**Reply by wobba** [reply]
If you are building the pages programmatically, then you are adding the web parts programmatically as well. When you add the web part you can set the JSON to your needs. Or change it after added to the page. All depends how your current code/script is is :)

**Reply by Niax01** [reply]
I am creating the pages from a  page template, where the web parts already are added with all the settings needed, except for the path, that needs to be changed. So my script uses an Excel-file to get the properties, like Title, FileName and a custom columns metadata value. So the only part left is how to update/change the path in the query template.
I have the JSON of the web part as an output to a txt-file, but I am not sure yet if it's in the right format, or if I need the whole JSON or if it's possible to only change the query template part?
This is all probably very easy once you know how to, I just need to get there :)

**Reply by Niax01** [reply]
Hi @wobba I actually got it working, thank you so much for your help!

/Nicklas
